### PR TITLE
Validate UseWorkItemFilters names against registered tasks at worker build time

### PR DESCRIPTION
**Pull request by YunchuWang**
# Summary
## What changed?
- Added `DurableTaskWorkerWorkItemFiltersValidator` (`IValidateOptions<DurableTaskWorkerWorkItemFilters>`) that cross-checks every orchestration / activity / entity name in the configured filters against the worker's `DurableTaskRegistry`.
- `UseWorkItemFilters(builder, filters)` now auto-registers this validator (only when explicit, non-empty filters are supplied) so customers do not have to invoke validation themselves. The validator runs once after every `PostConfigure` callback, so any later overwrites are still validated against the final state.
- Filter-name lookups use the registry's dictionary keys (`TaskName`, `OrdinalIgnoreCase`) so case differences and entity-name normalization match the runtime behavior.
- Failure message lists every unknown name grouped by Orchestrations / Activities / Entities so customers can fix all mismatches in one shot.
- CHANGELOG entry + XML doc updated to call out the new behavior.
- Two pre-existing tests that previously configured filters without registering the matching tasks were updated to register valid tasks; the behaviors they verify (filter propagation and explicit-overrides-auto-gen) are preserved.

## Why is this change needed?
When a customer calls `UseWorkItemFilters(...)` with a name that is not registered with the worker (typo, removed task, copy/paste mistake, etc.), the worker would previously start successfully and then silently wait for work items it can never handle — orchestrations / activities / entities targeting those names would simply never run, with no error surfaced. Failing fast at worker startup with a precise, actionable `OptionsValidationException` makes this misconfiguration immediately diagnosable.

## Issues / work items
- Resolves #
- Related #

---

# Project checklist
- [x] Release notes are not required for the next release
  - [x] Otherwise: Notes added to `release_notes.md` (entry added to `CHANGELOG.md` under `Unreleased`)
- [x] Backport is not required
  - [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
- [x] All required tests have been added/updated (unit tests, E2E tests)
- [x] Breaking change?
  - Behavior change only: previously-silent misconfiguration in `UseWorkItemFilters(filters)` now throws `OptionsValidationException` at worker startup. Existing callers whose filter names match registered tasks are unaffected.
  - Impact: Workers that were silently mis-configured (filtering for unregistered names) will now fail to start.
  - Migration guidance: Either register the missing task with the worker or remove the unknown name from the filters; the exception message lists every unknown name grouped by category.
---

# AI-assisted code disclosure (required)
## Was an AI tool used? (select one)
- [ ] No
- [ ] Yes, AI helped write parts of this PR (e.g., GitHub Copilot)
- [x] Yes, an AI agent generated most of this PR

If AI was used:
- Tool(s): GitHub Copilot (Claude)
- AI-assisted areas/files:
  - `src/Worker/Core/DependencyInjection/DurableTaskWorkerWorkItemFiltersValidator.cs`
  - `src/Worker/Core/DependencyInjection/DurableTaskWorkerBuilderExtensions.cs`
  - `test/Worker/Core.Tests/DependencyInjection/UseWorkItemFiltersTests.cs`
  - `test/Worker/Grpc.Tests/DurableTaskWorkerWorkItemFiltersExtensionTests.cs`
  - `CHANGELOG.md`
- What you changed after AI output: Reviewed the validator wiring (post-configure ordering, single-registration, opt-in only when explicit non-empty filters are supplied), confirmed registry-key comparison semantics match runtime behavior, tightened the error-message grouping, and updated two existing tests so they exercise valid task registrations.

AI verification (required if AI was used):
- [x] I understand the code and can explain it
- [x] I verified referenced APIs/types exist and are correct
- [x] I reviewed edge cases/failure paths (timeouts, retries, cancellation, exceptions)
- [x] I reviewed concurrency/async behavior
- [x] I checked for unintended breaking or behavior changes

---

# Testing
## Automated tests
- Result: Passed — added coverage in `UseWorkItemFiltersTests` for: unknown orchestration / activity / entity names producing `OptionsValidationException` at startup, multi-category error grouping in the message, case-insensitive matching against registry keys, validator only registering for explicit non-empty filters, validator re-running after later `PostConfigure` overrides, and entity-name normalization matching the runtime path.

## Manual validation (only if runtime/behavior changed)
- N/A — behavior is exercised by automated tests.

---

# Notes for reviewers
- The validator is registered only when explicit, non-empty filters are supplied via `UseWorkItemFilters(builder, filters)` — auto-generated filters (the no-args overload) are intentionally not validated, since they are derived directly from the registry and cannot be inconsistent with it.
- The validator key/comparer matches `DurableTaskRegistry`'s dictionary semantics so the validation is exactly aligned with what the worker would dispatch.

### CODE REVIEW

File: src/Worker/Core/DependencyInjection/DurableTaskWorkerBuilderExtensions.cs [hunk 1]
```diff
@@ -170,6 +180,21 @@ public static IDurableTaskWorkerBuilder UseWorkItemFilters(this IDurableTaskWork
...
+        if (workItemFilters is not null
+            && (workItemFilters.Orchestrations.Count > 0
+                || workItemFilters.Activities.Count > 0
+                || workItemFilters.Entities.Count > 0))
+        {
+            builder.Services.AddSingleton<IValidateOptions<DurableTaskWorkerWorkItemFilters>>(sp =>
+                new DurableTaskWorkerWorkItemFiltersValidator(
+                    builder.Name,
+                    sp.GetRequiredService<IOptionsMonitor<DurableTaskRegistry>>()));
+        }
```
[the remark this answers is not in the document]
Comment (YunchuWang) [reply src/Worker/Core/DependencyInjection/DurableTaskWorkerBuilderExtensions.cs]: Good catch — fixed in 0e71490. `UseWorkItemFilters(filters)` now uses a per-builder-name marker (`WorkItemFiltersValidatorRegistrationMarker`) and skips registration if a marker for the same `builder.Name` is already present in `IServiceCollection`, so repeat calls register exactly one `IValidateOptions<DurableTaskWorkerWorkItemFilters>` per named worker. Locked in by two new tests: `WorkItemFilters_RepeatedExplicitCalls_RegisterSingleValidatorPerWorker` (asserts a single DI registration after three calls) and `WorkItemFilters_RepeatedExplicitInvalidCalls_ReportFailureExactlyOnce` (asserts `Failures` contains a single entry against the final state).

File: src/Worker/Core/DependencyInjection/DurableTaskWorkerWorkItemFiltersValidator.cs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+            // TaskName equality is OrdinalIgnoreCase, mirroring how registered keys are compared.
+            if (!isRegistered(name))
+            {
+                unknown.Add(name);
+            }
+        }
+
+        return unknown;
+    }
```
[the remark this answers is not in the document]
Comment (YunchuWang) [reply src/Worker/Core/DependencyInjection/DurableTaskWorkerWorkItemFiltersValidator.cs:100]: Fixed in 0e71490. `FindUnknown` now constructs the `TaskName` explicitly via `new TaskName(name)` before calling `isRegistered`, so it no longer depends on the implicit `string` → `TaskName` operator existing.

[the remark this answers is not in the document]
Comment (YunchuWang) [reply src/Worker/Core/DependencyInjection/DurableTaskWorkerWorkItemFiltersValidator.cs:100]: Fixed in 0e71490. FindUnknown now constructs the TaskName explicitly via new TaskName(name) before calling isRegistered, so it no longer depends on the implicit string -> TaskName operator existing.

File: test/Worker/Core.Tests/DependencyInjection/UseWorkItemFiltersTests.cs [hunk 3]
```diff
@@ -394,6 +402,323 @@ public void WorkItemFilters_NamedBuilders_HaveUniqueDefaultFilters_WhenExplicitl
...
+
+        // Act
+        ServiceProvider provider = services.BuildServiceProvider();
+        IOptionsMonitor<DurableTaskWorkerWorkItemFilters> filtersMonitor =
+            provider.GetRequiredService<IOptionsMonitor<DurableTaskWorkerWorkItemFilters>>();
+        Action act = () => filtersMonitor.Get("test");
+
+        // Assert
+        act.Should().NotThrow();
+    }
```
[the remark this answers is not in the document]
Comment (YunchuWang) [reply test/Worker/Core.Tests/DependencyInjection/UseWorkItemFiltersTests.cs:761]: Fixed in 0e71490. The test now also inspects the IServiceCollection and asserts there is no IValidateOptions<DurableTaskWorkerWorkItemFilters> descriptor at all when UseWorkItemFilters(null) is called, in addition to the existing NotThrow assertion. So the test now verifies the claim its name makes (no validator registered), rather than just observing the side effect of validation passing for empty filters.

File: src/Worker/Core/DependencyInjection/DurableTaskWorkerBuilderExtensions.cs [hunk 4]
```diff
@@ -170,6 +181,16 @@ public static IDurableTaskWorkerBuilder UseWorkItemFilters(this IDurableTaskWork
...
 
+        // Register a single global validator that fails fast at worker build time if any filter
+        // references a task name that isn't registered with the corresponding worker. The validator
+        // dispatches per named-options instance via the 'name' parameter that the Options framework
+        // passes to IValidateOptions.Validate, so a single registration covers every named worker.
+        // TryAddEnumerable de-duplicates by implementation type, making repeat calls idempotent
+        // across all UseWorkItemFilters invocations.
+        builder.Services.TryAddEnumerable(ServiceDescriptor.Singleton<
+            IValidateOptions<DurableTaskWorkerWorkItemFilters>,
+            DurableTaskWorkerWorkItemFiltersValidator>());
```
[the remark this answers is not in the document]
Comment (YunchuWang) [reply src/Worker/Core/DependencyInjection/DurableTaskWorkerBuilderExtensions.cs]: Good catch. Restored the gate in 5c401a2: TryAddEnumerable now only runs when workItemFilters is non-null AND has at least one orchestration/activity/entity entry. Added NullExplicitFilters_DoNotRegisterValidator and EmptyExplicitFilters_DoNotRegisterValidator as lock-in tests asserting the IServiceCollection contains no IValidateOptions registration in those cases.

File: src/Worker/Core/DependencyInjection/DurableTaskWorkerWorkItemFiltersValidator.cs [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+            && unknownEntities.Count == 0)
+        {
+            return ValidateOptionsResult.Success;
+        }
+
+        StringBuilder sb = new();
+        sb.Append("Cannot configure work item filters for worker '").Append(name)
+          .Append("': the following filter names do not match any registered task. ")
+          .Append("Register them on the worker (via AddTasks/AddOrchestrator/AddActivity/AddEntity) ")
+          .Append("or remove them from the filters.");
```
[the remark this answers is not in the document]
Comment (YunchuWang) [reply src/Worker/Core/DependencyInjection/DurableTaskWorkerWorkItemFiltersValidator.cs:71]: Fixed in 5c401a2: when name is null or string.Empty (default-name worker), the failure message now displays <default> instead of an empty quoted name. The actual name is still used to resolve the registry. Added DefaultNamedWorkerInvalidFilter_FailureMessageUsesDefaultPlaceholder as a lock-in test.

File: test/Worker/Core.Tests/DependencyInjection/UseWorkItemFiltersTests.cs [hunk 6]
```diff
@@ -394,6 +402,398 @@ public void WorkItemFilters_NamedBuilders_HaveUniqueDefaultFilters_WhenExplicitl
...
+        IOptionsMonitor<DurableTaskWorkerWorkItemFilters> filtersMonitor =
+            provider.GetRequiredService<IOptionsMonitor<DurableTaskWorkerWorkItemFilters>>();
+        DurableTaskWorkerWorkItemFilters actual = filtersMonitor.Get("test");
+
+        // Assert
+        actual.Orchestrations.Should().ContainSingle(o => o.Name == nameof(TestOrchestrator));
+    }
+
+    [Fact]
+    public void WorkItemFilters_InvalidExplicitOverwrittenByNull_DoesNotThrow()
```
Comment (bachuv) [inline test/Worker/Core.Tests/DependencyInjection/UseWorkItemFiltersTests.cs:611]: Should this test have the `act.Should().NotThrow();` check?

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,6 +1,7 @@
 # Changelog
 
 ## Unreleased
+- Validate explicit `UseWorkItemFilters(filters)` filter names against the worker's `DurableTaskRegistry`. Filters that reference an orchestration, activity, or entity name not registered with the worker now throw `OptionsValidationException` at worker startup instead of silently waiting for work items that will never arrive. No customer-side validation call is required.
```
Comment (bachuv) [inline CHANGELOG.md]: Can we add the PR link?

File: test/Worker/Core.Tests/DependencyInjection/UseWorkItemFiltersTests.cs (hunk 6 above)
Comment (YunchuWang) [reply test/Worker/Core.Tests/DependencyInjection/UseWorkItemFiltersTests.cs:611]: Fixed in 5447f98: `WorkItemFilters_InvalidExplicitOverwrittenByNull_DoesNotThrow` now uses an explicit `Action act = () => filtersMonitor.Get("test")` and asserts `act.Should().NotThrow()` before checking the cleared filter state.

File: CHANGELOG.md (hunk 7 above)
Comment (YunchuWang) [reply CHANGELOG.md]: Fixed in 5447f98: the Unreleased changelog entry now includes the PR link `([#719](https://github.com/microsoft/durabletask-dotnet/pull/719))`.

**Review by bachuv — APPROVED** [review]
